design/20210209.certificates.k8s.io-adoption.md

```diff
@@ -0,0 +1,292 @@
+---
+title: certificates.k8s.io Adoption
+authors:
+  - "@joshvanl"
+reviewers:
+  - "@joshvanl"
+approvers:
+  - "@munnerz"
+  - @jetstack/team-cert-manager
+editor: "@joshvanl"
+creation-date: 2021-02-09
+last-updated: 2021-02-17
+status: provisional
+---
+
+# certificates.k8s.io Adoption
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+- [Reasons for Approach](#reasons-for-approach)
+  * [Read all access to `CertificateSigningRequests`](#read-all-access-to-certificatesigningrequests)
+  * [Namespaced issuers referenced by `CertificateSigningRequest`](#namespaced-issuers-referenced-by-certificatesigningrequest)
+  * [Duration Field](#duration-field)
+  * [CA Field](#ca-field)
+  * [External Issuers](#external-issuers)
+- [Proposal](#proposal)
+  * [Signers](#signers)
+  * [Conditions](#conditions)
+  * [API Changes](#api-changes)
+  * [Upgrading](#upgrading)
+  * [Graduation](#graduation)
+<!-- /toc -->
+
+## Summary
+
+In Kubernetes v1.19 the
+[`CertificateSigningRequest`](https://github.com/kubernetes/api/blob/48bd8381a38a486f8b3cdf28cf7334a45b182f2e/certificates/v1/types.go#L41)
+resource graduated to `certificates.k8s.io/v1`. This makes the concept of
+requesting, signing, and consuming certificates in Kubernetes first class
+objects. cert-manager is well placed to serve this resource type as it becomes
+more popular in the wider community, whilst preserving all the features and
+extensions that cert-manager has to offer.
+
+
+## Motivation
+
+With the extensiveness of cert-manager’s ecosystem, it is well placed to manage
+the core Kubernetes resource type. This gives cert-manager the ability to
+integrate with components that it otherwise wouldn't have, without changes to the
+third-party project.
+
+Having the "request" resource managed inside the API server has a better
+security profile (i.e. `UserInfo` fields are managed by the API server rather
+than a configurable webhook,
+[see](https://github.com/jetstack/cert-manager/pull/3630)).
+
+Consumers of this API do not require cert-manager as a hard dependency.
+cert-manager can cause issues with order of operation scenarios during
+installation or upgrades. For example, the `CertificateRequest` resource needs
+to be installed and the webhook ready for consumers/clients of this API to make
+requests.
+
+Support for the `CertificateSigningRequest` resource in cert-manager is a
+stepping stone to migrating to this resource instead of the `CertificateRequest`
+resource, if the project does indeed want to migrate.
```

Comment:
I propose to add a bit of context as to what problem we are solving when we let projects migrate from custom resource to the built-in CSR resource:

```suggestion
Being able to rely on the built-in `CertificateSigningRequest` instead of
having to maintain the custom resource definition `CertificateRequest`
brings some benefits with regard to cluster operations. In some setups, the
cert-manager installation is handled by two separate teams:

- The cert-manager CRDs and webhook are installed and managed by the
  operations team, which allows the operations team to enfore cluster
  policies;
- The cert-manager controller, cainjector and acmesolver are installed and
  managed by the developer team.

By supporting the built-in `CertificateSigningRequest` resource, we
decrease the coupling between the operations team and the developer team.
Since the CSR resource is built-in, it does not need any CRD installation,
and the webhook for this resource is integrated to the apiserver.

Support for the `CertificateSigningRequest` resource in cert-manager is a
stepping stone to migrating to this resource instead of the `CertificateRequest`
resource, if the project does indeed want to migrate.
```

What do you think?